src/hazelcore/layout/HTypDoc.re

```diff
@@ -71,12 +53,28 @@ let rec mk = (~parenthesize=false, ~enforce_inline: bool, ty: HTyp.t): t => {
         ]),
         d2,
       ]);
-    | Prod(ty1, ty2) =>
-      let (d1, d2) =
-        mk_right_associative_operands(precedence_Prod, ty1, ty2);
-      hcats([d1, hcats([text(","), choices([linebreak(), space()])]), d2]);
+    | Prod([]) => text("()")
+    | Prod([head, ...tail]) =>
+      [
+        mk'(
+          ~parenthesize=HTyp.precedence(head) >= HTyp.precedence_Prod,
+          head,
+        ),
+        ...List.map(
+             ty =>
+               mk'(
+                 ~parenthesize=HTyp.precedence(ty) > HTyp.precedence_Prod,
```

Comment:
True. I think parsing would also be changed accordingly though right? I'm not really sure how to do that.

Reply:
I meant `,` is still parsed like a right-associative operator right? I think changing this to `>=` wouldn't directly cause any problems, because it would only generate extra parentheses, but I thought it would be nice to keep them in sync.
Anyway, I'll change it to `>=`.

Reply:
I changed the parsing in 0714710de73f3726c8140c4b36de90d8b4461e7f

Reply:
While working on this I also noticed the parenthesizing logic is duplicated in `HTypDoc` and `UHTyp`. I think it could be unified by using `contract` in `HTypDoc.mk`, but then you are pretty printing a `UHTyp`, for which `UHDoc` might be useful, but I don't really know anything about how all the layout stuff works.
What do you think about this?